### PR TITLE
Added a supercell builder.

### PR DESCRIPTION
**Pull request by psavery**

### CODE REVIEW

**Comment by psavery** [comment]
Here is a sample test.

Before supercell building:

O1Ti1
1
    4.04227659   0.00000000   0.00000000
    3.05273519   1.42155121   0.00000000
    3.54255818   1.83286254   2.40850502
O Ti
1 1
Direct
    0.00000000   0.75000000   0.75000000
    0.50000000   0.50000000   0.50000000


After supercell building:

O6 Ti6
1
    8.08455317   0.00000000   0.00000000
    3.05273519   1.42155121   0.00000000
   10.62767455   5.49858762   7.22551506
O Ti
6 6
Direct
    0.00000000   0.75000000   0.25000000
    0.00000000   0.75000000   0.58333333
    0.00000000   0.75000000   0.91666667
    0.50000000   0.75000000   0.25000000
    0.50000000   0.75000000   0.58333333
    0.50000000   0.75000000   0.91666667
    0.25000000   0.50000000   0.50000000
    0.75000000   0.50000000   0.16666667
    0.25000000   0.50000000   0.16666667
    0.75000000   0.50000000   0.50000000
    0.25000000   0.50000000   0.83333333
    0.75000000   0.50000000   0.83333333


Results are the same for Avogadro 1

**Comment by psavery** [comment]
Oh, and that's a 2x1x3 supercell by the way

File: avogadro/core/crystaltools.h [hunk 1]
```diff
@@ -102,6 +102,17 @@ class AVOGADROCORE_EXPORT CrystalTools
...
   /**
+   * Build a supercell by expanding upon the unit cell of @a molecule. It will
+   * only return false if the molecule does not have a unit cell or if a, b, or
+   * c is set to zero.
+   * @param a The number of units along lattice vector a for the supercell
+   * @param b The number of units along lattice vector b for the supercell
+   * @param c The number of units along lattice vector c for the supercell
+   * @return True on success, false on failure.
+   */
+  static bool buildSupercell(Molecule &molecule, uint a, uint b, uint c);
```
Comment (cryos) [inline avogadro/core/crystaltools.h]: I don't think uint is part of C++, http://stackoverflow.com/questions/3552094/c-uint-unsigned-int-int seems to agree on a quick search. We should use unsigned int, and if others are lurking use the standard type. Qt provides some with size guarantees I think.

File: avogadro/qtplugins/crystal/supercelldialog.cpp [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+  // If the user rejected, just return false
+  if (this->exec() == QDialog::Rejected) return false;
+
+  // Read the values
+  uint a = m_ui->aCellSpinBox->value();
+  uint b = m_ui->bCellSpinBox->value();
+  uint c = m_ui->cCellSpinBox->value();
+
+  // No need to do anything if all the values are one
+  if (a == 1 && b == 1 && c == 1) return true;
```
Comment (cryos) [inline avogadro/qtplugins/crystal/supercelldialog.cpp]: New line for statement

File: avogadro/qtplugins/crystal/supercelldialog.cpp [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+
+SupercellDialog::~SupercellDialog()
+{
+  delete m_ui;
+}
+
+bool SupercellDialog::buildSupercell(Avogadro::Core::Molecule &mol)
+{
+  // If the user rejected, just return false
+  if (this->exec() == QDialog::Rejected) return false;
```
Comment (cryos) [inline avogadro/qtplugins/crystal/supercelldialog.cpp]: New line for statement

File: avogadro/qtplugins/crystal/supercelldialog.h [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the License.
+
+******************************************************************************/
+
+#ifndef SUPERCELL_DIALOG_H
```
Comment (cryos) [inline avogadro/qtplugins/crystal/supercelldialog.h]: The header guards don't follow the general style in Avogadro

**Comment by cryos** [comment]
Looks good, I have been wanting to bring this back for ages. It seems to lack a call to reset the bounds, so had clipping problems here, but otherwise looked good from a usability perspective.

**Comment by psavery** [comment]
I fixed all the things you mentioned except for the clipping problem.

It appears that the clipping problem is not just for building supercells (so maybe I should send in a separate pull request for that). If I scale the cell volume a lot or edit the unit cell and make it larger, it also gets clipped. However, if I save it and then reload it, it is no longer clipped.

I might need a little bit of help figuring out how to fix it. I think CrystalScene::process() in the following file is what draws the updated unit cell:

https://github.com/OpenChemistry/avogadrolibs/blob/master/avogadro/qtplugins/crystal/crystalscene.cpp

However, I don't see in Molecule or GroupNode a simple way to reset the bounds.

Is LineStripGeometry::addLineStrip() supposed to reset the bounds?

**Comment by psavery** [comment]
I updated this so that there is no longer a merge conflict. I also sent in a separate pull request with a fix to the clipping problem.

**Comment by cryos** [comment]
Looks good, thanks!

**Pull request merged**